revise style guidelines for capitalization (#24645)
kubernetes/website: content/en/docs/contribute/style/style-guide.md

### Use upper camel case for API objects

> When you refer specifically to interacting with an API object, use [UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case), also known as Pascal Case. You may see different capitalization, such as "configMap", in the  [API Reference](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/). It's better to use UpperCamelCase and call that "ConfigMap" instead when writing general documentation. 

Reviewer: Pascal case. (lowercase "c")

When writing general documentation, it's better to use upper camel case, calling it "ConfigMap" instead.

Author: fixed, thanks!

> When you refer specifically to interacting with an API object, use [UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case), also known as Pascal case. You may see different capitalization, such as "configMap", in the  [API Reference](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/). When writing general documentation, it's better to use upper camel case, calling it "ConfigMap" instead.

Reviewer 2: These paragraphs are clear. Since the reference section has changed, I suggest to clean up the link to the API reference.
For example:
```
[API Reference](/docs/reference/kubernetes-api/)
```

Author: 👍will fix

When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).

> You may use the word "resource", "API", or "object" to clarify a Kubernetes resource type in a sentence. Note that Kubernetes resource names are capitalized. However, capitalization alone is not a reliable way to identify Kubernetes resource types in a sentence. Review the related guidance on [Code Style](#code-style-inline-code-commands-api-objects)

Reviewer 2: ```suggestion
You may use the word "resource", "API", or "object" to clarify a Kubernetes resource type in a sentence. 
```

Author: fixed, thanks!

Don't split the API object name into separate words. For example, use
PodTemplateList, not Pod Template List.

> The following examples focus on capitalization. Review the related guidance on [Code Style](#code-style-inline-code) for more information on formatting API objects. 

Reviewer 2: I don't think you need to describe what to expect in the following table. How about removing line 56?

Author: I added that line to offer an hint as to why code style appeared in the table, so I think it has a purpose.

> {{< table caption = "Do and Don't - Use code style for API objects" >}}

Reviewer 2: I don't completely understand this table caption change.
How about:
Use Pascal case for API objects OR
Use Pascal case and code style for API objects

Author: this was a mistake on my part 😂 your suggestion is what I meant.

> {{< table caption = "Do and Don't - Use Pascal case for API objects" >}}

Reviewer 2: nit: "upper camel case" vs. Pascal case?

The following examples focus on capitalization. Review the related guidance on [Code Style](#code-style-inline-code) for more information on formatting API objects. 

{{< table caption = "Do and Don't - API objects" >}}
{{< table caption = "Do and Don't - Use Pascal case for API objects" >}}
Do | Don't
:--| :-----
The pod has two containers. | The Pod has two containers.

> The HorizontalPodAutoscaler is responsible for ... | The HorizontalPodAutoscaler object is responsible for ...

Reviewer 3: One thing to watch out for: you can find a volume mentioned in Kubernetes in several places, but AIUI _not_ as a resource in the HTTP API. I'm wary in case someone reads the style guide and ends up confused about how storage works.

> The HorizontalPodAutoscaler is responsible for ... | The HorizontalPodAutoscaler object is responsible for ...

Reviewer 2: OK @sftim. Can you provide a concrete example?

> The HorizontalPodAutoscaler is responsible for ... | The HorizontalPodAutoscaler object is responsible for ...

Reviewer 2: Could remove the "Volume" example. The examples I provided are "tests" of the recommendations.

> The HorizontalPodAutoscaler is responsible for ... | The HorizontalPodAutoscaler object is responsible for ...

Reviewer 2: storage volume, volume plugin, a volume or logical drive, volume driver

> The HorizontalPodAutoscaler is responsible for ... | The HorizontalPodAutoscaler object is responsible for ...

Author: I agree on moving this forward or closing it. 

I'm thinking I'll bring this up again at the next sig docs meeting to ask for synchronous feedback. I'll make a powerpoint slide with the new examples. 

Regarding the volume example, I attempted to literally copy it from the API reference. 

See, for example, 
- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#volume-v1-core
- this page defines a volume API resource that includes a `hostpath` field.

- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podspec-v1-core
- podspec includes a field that is an array of volumes

> {{< table caption = "Do and Don't - Upper camel case for API objects" >}}

Reviewer 2: ```suggestion
{{< table caption = "Do and Don't - API objects" >}}
```

Author: done, thanks!

> There is a HostPath field in the Volume resource. | There is a hostpath in the volume. 

Reviewer 4: The "HostPath" *must* be spelled as "hostPath" because it is about doc accuracy rather than style. There is no field named "HostPath" in a Volume resource. That field is called "hostPath". 

The confusion may be caused by the fact that in the Go source code, the field is named "HostPath", with "H" capitalized or else the field cannot be accessed out of the package.

> There is a HostPath field in the Volume resource. | There is a hostpath in the volume. 

Reviewer 2: 👋 @geoffcline , Would you apply @tengqm 's comments OR change the example?
The `Volume` resource contains a `hostPath` field. | The volume resource contains a hostPath field.

Author: thank you! I changed the example.

> The `Volume` resource contains a `hostPath` field. | The volume resource contains a hostPath field.

Reviewer 3: [Volumes](https://kubernetes.io/docs/concepts/storage/volumes/) exist in the Kubernetes API, but not as resources (see [API terminology](https://kubernetes.io/docs/reference/using-api/api-concepts/#standard-api-terminology)).

To avoid using a misleading example, I suggest using PersistentVolume rather than Volume, and the `volumeMode` field; otherwise, keep the two sentences the same.

Author: resolved, thanks!

> Every `ConfigMap` is part of a namespace | Every configMap is part of a namespace. 

Reviewer 2: nit: Add a period at end of "Do" sentence.

Author: good catch!

> For managing confidential data, consider using the Secret API. | For managing confidential data, consider using a secret.

Reviewer 4: Why not wrap "Secret" with backtiqs?
Just trying to make sure the style is consistent... we are surrounding "ConfigMap" with backtiqs on line 63.

Looks good to me other than this "inconsistency".

Author: I want to illustrate different options in the do/don't table, such as backticks or saying "the Secret API".

Reviewer 4: Okay I see. That makes sense.

> For managing confidential data, consider using the Secret API. | For managing confidential data, consider using a secret.

Reviewer 2: :wave: @geoffcline .

Does the `Secret` example represent this capitalization case:
`When you are generally discussing an API object, use [sentence-style capitalization]`

Personal opinion: I am not convinced the site needs to format API resources and definitions as code.
To be consistent, any example in this guide should format API objects with the recommended style.
Otherwise, if there are remaining questions about formatting, raise an issue. Lines 62-64 add new content and should follow the style guide. Lines 62 and 63 display an API resource, definition, and field name in code format.

Would this example fit with the current style and capitalization recommendations?
```
For managing confidential data, consider using a `Secret`. | For managing confidential data, consider using a secret.
```

Author: Thank you for reviewing again. 

Did you see my discussion on this issue with Qiming? https://github.com/kubernetes/website/pull/24645#discussion_r540642351

I don't think our style guide *mandates* code style yet, so I wanted to illustrate the different options.

> Every ConfigMap is part of a Namespace | Every configMap object is part of a Namespace object. 

Reviewer 2: nit?:
Every `ConfigMap` is part of a namespace | Every configMap object is part of a Namespace object.

If it applies, I think it is reasonable to show an API term with code styling.

Author: I'm really trying to keep the do/don't section focused to illustrate one concept at at time. I hope users will read the preceding paragraph and review the code style examples also.

Author: Although, after considering, I think it makes sense to have one "do" example include code style.

> For managing confidential data, consider using the Secret API object. | For managing confidential data, consider using a secret.

Reviewer 2: nit:
For managing confidential data, consider using the Secret API. | For managing confidential data, consider using a secret.
OR
For managing confidential data, consider using a `Secret`. | For managing confidential data, consider using a secret.
OR
For managing confidential data, consider using the Secret resource. | For managing confidential data, consider using a secret.

Author: Thanks for this suggestion. I think the revised examples illustrate multiple ways to distinguish API objects for clear writing, such as code style, capitalization, and adding "API" or "resource" terms.

When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).

You may use the word "resource", "API", or "object" to clarify a Kubernetes resource type in a sentence. Note that Kubernetes resource names are capitalized. However, capitalization alone is not a reliable way to identify Kubernetes resource types in a sentence. Review the related guidance on [Code Style](#code-style-inline-code-commands-api-objects)

Don't split the API object name into separate words. For example, use
PodTemplateList, not Pod Template List.

The following examples focus on capitalization. Review the related guidance on [Code Style](#code-style-inline-code-commands-api-objects) for more information on formatting API objects. 

{{< table caption = "Do and Don't - Upper camel case for API objects" >}}
Do | Don't
:--| :-----
A PodList is a list of pods. | A Pod List is a list of pods.

> There is a HostPath field is the Volume resource. | Hostpath is a type of volume.

Reviewer 4: ```suggestion
There is a `hostPath` field in the Volume resource. | Hostpath is a type of volume.
```

Reviewer 2: Sorry for the nit, but I am not sure about the "Do not" example because you begin a sentence
with the volume type name. What is the name of the volume type? "hostPath" is the volume plugin type?

Author: I corrected is/in, and harmonized the do/don't examples. 

> There is a HostPath field in the Volume resource. | There is a hostpath field in the volume resource.

When you refer specifically to interacting with an API object, use [UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case), also known as Pascal Case. When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).

> You may use the word "resource", "API", or "object" to clarify a Kubernetes resource type in a sentence. Note that Kubernetes resource names are capitalized. However, capitalization alone is not a reliable way to identify Kubernetes resource types in a sentence. 

Reviewer 2: Hi @geoffcline .
After reading through the changes, I think it would be worthwhile to focus on simplicity
and try to make small changes to move this forward.
Any guidance that is not explicit requires more thought to put into action.
I think the main point of this change is:
```You may use the word "resource", "API", or "object" to clarify a Kubernetes resource type in a sentence.```
I think you could remove the other sentences about capitalization and change this sentence as you see fit.
I suggest to remove the example sentences and instead push these examples into the tables (without
the dialog about what is right or wrong). Show by example.

Author: I sincerely appreciate your guidance on making small changes to move this forward :)

> The sentence "However, pods may have multiple volume mounts" describes general concepts, and uses lower case. In contrast, the sentence "The Pod object includes a Volumes field" is talking about API objects, and uses UpperCamelCase. You may also use [Code Style](#code-style) to distinguish API objects. For example, "The `Pod` object includes a `volumes` field."

Reviewer 3: ```suggestion
The sentence "However, pods may have multiple volume mounts" describes general concepts, and uses lower case. In contrast, the sentence "The Pod object includes a volumes field" is talking about the Kubernetes API, and uses UpperCamelCase. To distinguish fields and resource names from natural language, use [Code Style](#code-style) whereever there is any risk of confusion. For example, "The `Pod` object includes a `.volumes` field."
```

Reviewer 4: damn

Reviewer 2: ?

Reviewer 4: - Since when we agreed to include a "." in the field name? Is this an improvement or an intent to drive this astray?
- Is "whereever" a newly invented English word?
- The proposed change is about "API objects", which is very clear. Why are we trying to mess this with "Kubernetes API"?

Reviewer 2: I would leave out the information starting on line 51.

Author: I am going with @kbhawkey 's suggestion to remove line 51 and keep this PR focused. I will save a copy of the proposed changes for further consideration.

> Capitalization varies in the Kubernetes API. However, API objects are documented in UpperCamelCase. For example, you may see "configMap" in the [API reference](https://kubernetes.io/docs/reference/). It's fine to call that "ConfigMap" instead when writing general documentation. 

Reviewer 3: ```suggestion
Capitalization varies in the Kubernetes API. However, the `.kind` of an API object is commonly written in UpperCamelCase. For example, you may see `configMap` in some areas of the [API Reference](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/). It's better to call that type of object a `ConfigMap` when writing general documentation.
```

Reviewer 2: Line 53, Again, the sentence is about capitalization/upper camel case. If you feel that this content
clarifies the use of upper camel case in the documentation, you could add it after the first
sentence.

Reviewer 2: I am unsure whether this style information is warranted (line 53).

Don't split the API object name into separate words. For example, use
PodTemplateList, not Pod Template List.

{{< table caption = "Do and Don't - API objects" >}}
Do | Don't
:--| :-----
A PodList is a list of pods. | A Pod List is a list of pods.
HostPath is a type of Volume. | Hostpath is a type of volume.

> `HostPath` is a type of Volume. | Hostpath is a type of volume.

Reviewer 3: I'd write:
```suggestion
`HostPath` is a type of volume. | Hostpath is a type of volume.
```

and if I saw a page that read “`HostPath` is a type of Volume”, I'd probably raise an issue about fixing that. People never `kubectl create volume` whereas they might `kubectl create deployment`, so “Deployment” gets UpperCamelCase but “volume” usually doesn't. That's my rule of thumb for deciding what capitalization to apply: can you interact with it via `kubectl`? There are exceptions but they are rare.

Reviewer 2: :+1:

PodTemplateList, not Pod Template List.

Refer to API objects without saying "object," unless omitting "object"
leads to an awkward construction.
There is a subtle distinction between formal API objects and general concepts. Use "object" where appropriate to clarify the meaning of terms. Calling out a "Pod object" may be appropriate when moving from general concepts to specific API objects. For example, "Pods may have multiple volume mounts. The Pod object includes a Volumes field."

Do not use "object" repetitively, such as where it's clear a term is an API object. 

Note that capitalization is a weak way to signal information to the reader. This information may be lost entirely due to assistive technology, differences in visual abilities, or translation. 

{{< table caption = "Do and Don't - API objects" >}}
Do | Don't
:--| :-----
The pod has two containers. | The Pod has two containers.
The HorizontalPodAutoscaler is responsible for ... | The HorizontalPodAutoscaler object is responsible for ...
A PodList is a list of pods. | A Pod List is a list of pods.
The two ContainerPorts ... | The two ContainerPort objects ...

> The two ContainerStateTerminated objects ... | The two ContainerStateTerminateds ...

Reviewer 4: Please propose alternatives to these.

Author: I added two more (imperfect) examples.

Don't split the API object name into separate words. For example, use
PodTemplateList, not Pod Template List.

> There is a subtle distinction between formal API objects and general concepts. Use "object" where appropriate to clarify the meaning of terms. Calling out a "Pod object" may be appropriate when moving from general concepts to specific API objects. For example, "Pods may have multiple volume mounts. The Pod object includes a Volumes field."

Reviewer 3: Perhaps:
```suggestion
There is a subtle distinction between formal API objects and general concepts. Use "object" where appropriate to clarify the meaning of terms. Calling out a "`Po` object" may be appropriate when moving from general concepts to specific API objects. For example, "pods may have multiple volume mounts. The `Pod` object includes a `volumes` field."
```
?

> Do not use "object" repetitively, such as where it's clear a term is an API object. 

Reviewer 2: Hello @geoffcline , Some feedback:
Good idea to relax the use of "object" with resource types. If in doubt, I think it is a good idea
to log an issue to propose a change (not everyone is able to attend a meeting).
Note: The current guidance was probably added to curb overuse.

nit: I am tempted to drop line 54.
Does this suggest that the content should introduce an API term with "object"
and then drop the use of "object" with further instances of the term within a page?
I don't know whether the guide needs to go into this level of detail on the usage of "object".
What do you think about adding:
`You may use the word "resource", "API", or "object" to clarify a Kubernetes resource type in a sentence.
Note that Kubernetes resource names are capitalized. However, capitalization alone is not a reliable
way to identify Kubernetes resource types in a sentence`.

Regarding the statement about "weak signal and lost due to ...", this seems to be referring to 
the same possible issue of using code formatting for resource types?
I can't tell whether this type of visual hint to the reader is helpful. What do you think?

Author: I like your wording on clarifying resource types, thanks. 

I agree about the similarities between capitalization and code style. I do think code style sends a bit clearer signal, and is a better way to indicate something is a string used in code (and thus shouldn't be translated). 

I lean towards offering people more options, but I understand we may want the style guidelines to be more definitive.

I'm optimistic my next draft will address these questions.

Do not use "object" repetitively, such as where it's clear a term is an API object. 

Note that capitalization is a weak way to signal information to the reader. This information may be lost entirely due to assistive technology, differences in visual abilities, or translation. 

{{< table caption = "Do and Don't - API objects" >}}
Do | Don't
:--| :-----
The pod has two containers. | The Pod has two containers.

> The HorizontalPodAutoscaler is responsible for ... | The HorizontalPodAutoscaler object is responsible for ...

Reviewer 3: I suggest
```suggestion
The `HorizontalPodAutoscaler` is responsible for ... | The HorizontalPodAutoscaler object is responsible for ...
```
(and keep the line)

Author: I had purposely left out code style from the do/don't table, because I was wary of trying to illustrate two concepts at once. 

I'm leaning toward using code style in the table, but I will add a link to the code style guidelines in the preceding paragraph.

Reviewer: Geoff has the instinct of correct. In the present example, "HorizontalPodAutoscaler" should be set only in code (in all examples) or not in code at all. Mixing them up only confuses two issues, and the reader. A brief spleen-dump concerning code formatting follows. 

Once upon a time, monospace type was used to indicate command-line interactions and code snippets, full stop. Your guide now recommends it for paths, command tools, and component names. The present example adds Kubernetes objects to this list. This is a mistake: It feels like you're making a bright distinction, because you already know what you're trying to tell the reader, and putting it in monospace makes it stand out and look like documentation. In the present example, Pascal case already conveys quite clearly that the word HorizontalPodAutoscaler is an object, or at least is not an ordinary word (are objects considered components? Who adjudicates this?) Setting it in monospace adds no new meaning and burdens your writers and readers with another exceptional rule. As soon as someone copies and pastes the passage, the monospace signifier vanishes, leaving you where you started. 

As @tengqm pointed out earlier, capitalization is a weak way of conveying information. Modifying typesetting is even weaker, especially when rules—however well-intentioned they may be—require its use to proliferate. (If every word is bold, no word is bold...)

Author: this is a good point about copy/paste. 

I don't want to wade too deep into the code style debate here. The new draft page provides flexibility, and I think is a good next step.

Reviewer 3: > I don't want to wade too deep into the code style debate here.

_This_ PR should IMO either revise the code style recommendations or fit in with them. I'm not so keen on adding recommendations that conflict with other parts of the same page.

The easy option is to follow the current guidance on code style, and then if this merges come back and tweak it. It's the same idea as keeping PRs small that change source code: many small PRs that make incremental improvements, rather than one large change set that aims for perfection.

Reviewer 2: I agree that the style guide now offers more styling options for API objects.

* Do you need to show code styling in this section or could you state that the examples focus
on capitalization and usage of general terms?
* Is code styling of API objects optional? Is it acceptable to use code styling of a Kubernetes
resource term that also adds "resource" or "object" after the term? Probably not.

Again, this section comments on how the API resource names/definitions are capitalized (which 
matches how the name is used in the specification). This section also adds that it is acceptable to
add the word "object" after the term. Thirdly, the section adds that at times it is acceptable to use a general term (with sentence case) to refer to Kubernetes related software objects such as pod, node, secret, service, container.

> A PodList is a list of pods. | A Pod List is a list of pods.

Reviewer 3: Maybe delete this line, or clarify why it's right to retain it.

Author: I think it should be retained to illustrate 

> Don't split the API object name into separate words. For example, use
> PodTemplateList, not Pod Template List.

Reviewer 3: Ok, makes sense.

> HostPath is a type of Volume. | Hostpath is a type of volume.

Reviewer 3: ```suggestion
`VolumeAttachment` objects are non-namespaced. | VolumeAttachments aren't namespaced.
For managing confidential data, consider using a `Secret`. | For managing confidential data, consider using a secret.
```

> The Volume object is tied to a Pod. | The volume is tied to a pod.

Reviewer 3: ```suggestion
Every `ConfigMap` is part of a `Namespace`. |  Every config map object  is part of a Namespace object.
```

> {{< /table >}}

Author: I'm open to feedback on the revised do/don't table

Reviewer 4: The example on line 63 is a little confusing to me, a non-native English speaker. I'm okay to this sentence:

> The Volume object is referenced by a Pod.

Because in this context, saying "The Volume is referenced by a Pod" instead is not acceptable, because we are obviously talking about the Volume API object.

As a second example, the following may be NOT okay:

> That Volume object is mounted when the Pod is created.

The "object" should be deleted here because we are talking about the storage volume represented by a Volume.

You example is confusing to me because the meaning of "tied to" is vague.

Author: It's valuable information that the proposed example is unclear to non-native English speakers. I am revising the examples based on Tim's suggestions, and I think that will solve this problem. Thanks!

## Inline code formatting

### Use code style for inline code, commands, and API objects {#code-style-inline-code}

> 

Reviewer 2: You could update line 123 so that the table caption matches the section heading from line 118:
```{{< table caption = "Do and Don't - Use code style for inline code and commands and API objects" >}}```

Author: will do

> ### Use code style for inline code, commands, and API objects {#code-style-inline-code-commands-api-objects}

Reviewer 2: I anchor text is wordy. The table caption for this section needs updating.

Author: I fixed the anchor text, and searched for references to the anchor. 

BTW, do you have a suggested tool to search for broken #links in a local repo?

I'm not sure what you mean by the table caption needing updating. I think this is the only specific guidance, https://v1-17.docs.kubernetes.io/docs/contribute/style/style-guide/#tables

I attempted to conform the table captions to the prevailing style in the document.  

Thanks again!

## Inline code formatting

> ### Use code style for inline code, commands, and API objects {#code-style}

Reviewer 2: nit: I would make the anchor text more specific. There are a couple of `code style` related sections after this section.
Possibly:
`{#code-style-inline-code-commands-api-objects}`
